In this notebook, we will process the dataset loaded form Kaggle to be useful for the e-commerce catalog. 

The selected dataset is "__Ecommerce Product Dataset__" that can be downloaded from Kaggle. You can access the dataset from [here](https://www.kaggle.com/datasets/aaditshukla/flipkart-fasion-products-dataset?resource=download)

# Loading packages

In [4]:
import pandas as pd 
import json

# Data Loading 

Let's load the data that has been stored locally:

In [5]:
# read the excel file and load it in a dataframe
df = pd.read_excel('output.xlsx', index_col=0)
df.head()

_id actual_price  average_rating brand  \
0  fa8e22d6-c0b6-5229-bb9e-ad52eda39a0a        2,999             3.9  York   
1  893e6980-f2a0-531f-b056-34dd63fe912c        1,499             3.9  York   
2  eb4c8eab-8206-59d0-bcd1-a724d96bf74f        2,999             3.9  York   
3  3f3f97bb-5faf-57df-a9ff-1af24e2b1045        2,999             3.9  York   
4  750caa3d-6264-53ca-8ce1-94118a1d8951        2,999             3.9  York   

                   category            crawled_at description  \
0  Clothing and Accessories  02/10/2021, 20:11:51     69% off   
1  Clothing and Accessories  02/10/2021, 20:11:52     66% off   
2  Clothing and Accessories  02/10/2021, 20:11:52     68% off   
3  Clothing and Accessories  02/10/2021, 20:11:53     69% off   
4  Clothing and Accessories  02/10/2021, 20:11:53     68% off   

                                            discount  \
0  Yorker trackpants made from 100% rich combed c...   
1  Yorker trackpants made from 100% rich combed c...   
2  Yorker trackpants made from 100% rich combed c...   
3  Yorker trackpants made from 100% rich combed c...   
4  Yorker trackpants made from 100% rich combed c...   

                                              images  out_of_stock  \
0  ['https://rukminim1.flixcart.com/image/128/128...         False   
1  ['https://rukminim1.flixcart.com/image/128/128...         False   
2  ['https://rukminim1.flixcart.com/image/128/128...         False   
3  ['https://rukminim1.flixcart.com/image/128/128...         False   
4  ['https://rukminim1.flixcart.com/image/128/128...         False   

                pid                                    product_details  \
0  TKPFCZ9EA7H5FYZH  [{'Style Code': '1005COMBO2'}, {'Closure': 'El...   
1  TKPFCZ9EJZV2UVRZ  [{'Style Code': '1005BLUE'}, {'Closure': 'Draw...   
2  TKPFCZ9EHFCY5Z4Y  [{'Style Code': '1005COMBO4'}, {'Closure': 'El...   
3  TKPFCZ9ESZZ7YWEF  [{'Style Code': '1005COMBO3'}, {'Closure': 'El...   
4  TKPFCZ9EVXKBSUD7  [{'Style Code': '1005COMBO1'}, {'Closure': 'Dr...   

              seller selling_price sub_category  \
0  Shyam Enterprises           921   Bottomwear   
1  Shyam Enterprises           499   Bottomwear   
2  Shyam Enterprises           931   Bottomwear   
3  Shyam Enterprises           911   Bottomwear   
4  Shyam Enterprises           943   Bottomwear   

                               title  \
0   Solid Men Multicolor Track Pants   
1         Solid Men Blue Track Pants   
2   Solid Men Multicolor Track Pants   
3   Solid Men Multicolor Track Pants   
4  Solid Men Brown, Grey Track Pants   

                                                 url  
0  https://www.flipkart.com/yorker-solid-men-mult...  
1  https://www.flipkart.com/yorker-solid-men-blue...  
2  https://www.flipkart.com/yorker-solid-men-mult...  
3  https://www.flipkart.com/yorker-solid-men-mult...  
4  https://www.flipkart.com/yorker-solid-men-brow...

Let's check for the number of the records and features in the data. 

In [6]:
df.shape

(30000, 17)

In this dataset, we have 30,000 records and 17 featues for each of them. 

As in our catalog, the interesting field for us are only 'product name', 'price', 'category', 'description' and 'rating' we will only filter out those columns from data. 

In [7]:
# filter out the interesting fields
df_subset = df[['title', 'actual_price', 'category', 'discount', 'average_rating']]
df_subset.head()

title actual_price                  category  \
0   Solid Men Multicolor Track Pants        2,999  Clothing and Accessories   
1         Solid Men Blue Track Pants        1,499  Clothing and Accessories   
2   Solid Men Multicolor Track Pants        2,999  Clothing and Accessories   
3   Solid Men Multicolor Track Pants        2,999  Clothing and Accessories   
4  Solid Men Brown, Grey Track Pants        2,999  Clothing and Accessories   

                                            discount  average_rating  
0  Yorker trackpants made from 100% rich combed c...             3.9  
1  Yorker trackpants made from 100% rich combed c...             3.9  
2  Yorker trackpants made from 100% rich combed c...             3.9  
3  Yorker trackpants made from 100% rich combed c...             3.9  
4  Yorker trackpants made from 100% rich combed c...             3.9

As observed in the data, there are some mismatching between the columns names and the ones desired for our project. So we will also change the columns that are desirable for us. 


In [8]:
# change the name of the columns in order 
df_subset.columns = ['product_name', 'price', 'category', 'description', 'rating']

df_subset.head()

product_name  price                  category  \
0   Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
1         Solid Men Blue Track Pants  1,499  Clothing and Accessories   
2   Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
3   Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
4  Solid Men Brown, Grey Track Pants  2,999  Clothing and Accessories   

                                         description  rating  
0  Yorker trackpants made from 100% rich combed c...     3.9  
1  Yorker trackpants made from 100% rich combed c...     3.9  
2  Yorker trackpants made from 100% rich combed c...     3.9  
3  Yorker trackpants made from 100% rich combed c...     3.9  
4  Yorker trackpants made from 100% rich combed c...     3.9

# Data Cleaning

In order to work with a clean data, we will check for any missing values and duplicated records in our data. 

## Missing values 

In [9]:
# check for the number of the missing values 
df_subset.isnull().sum()

product_name        0
price             863
category            0
description     11983
rating           2446
dtype: int64

As it can be seen from the data, there are som missing values existed in our dataset. In here we can take two approaches: 

1. Accepting the data as it is
2. Dropping the records that have missing values. 

We can select the second approach, as for building a trustworthy website and and retail we should be able to provide all of the important and necessary information to the customer. In real-life, we can ask the admin and other responsible to address these missed values and fill the in with correct data. 

In [10]:
# drop the records that have missing values 
df_clean = df_subset.dropna()
df_clean

product_name  price                  category  \
0       Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
1             Solid Men Blue Track Pants  1,499  Clothing and Accessories   
2       Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
3       Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
4      Solid Men Brown, Grey Track Pants  2,999  Clothing and Accessories   
...                                  ...    ...                       ...   
29985       Full Sleeve Solid Men Jacket  2,999  Clothing and Accessories   
29989     Striped Men Black Basic Shorts    999  Clothing and Accessories   
29990      Solid Men Grey Bermuda Shorts    999  Clothing and Accessories   
29991      Solid Men Grey Bermuda Shorts    999  Clothing and Accessories   
29992       Solid Men Black Night Shorts    799  Clothing and Accessories   

                                             description  rating  
0      Yorker trackpants made from 100% rich combed c...     3.9  
1      Yorker trackpants made from 100% rich combed c...     3.9  
2      Yorker trackpants made from 100% rich combed c...     3.9  
3      Yorker trackpants made from 100% rich combed c...     3.9  
4      Yorker trackpants made from 100% rich combed c...     3.9  
...                                                  ...     ...  
29985  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29989  OKANE endeavors to continue to be a leader in ...     3.8  
29990  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29991  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29992  ‘OKANE’ endeavors to continue to be a leader i...     3.8  

[16082 rows x 5 columns]

We can observe that, the number of the records have been decreased from '30,000' records to '16,082' records. 

## Duplicated records 

In a retail store that doesn't often happens that the same product with exact same information be appeared. In this case there should not be any duplicated records in the data. We will check if there is any duplicated records in the data and drop them in case there is any. 

In [11]:
# check for the number of the duplicated records in the data 
df_clean.duplicated().sum()

np.int64(5970)

It seems that there are also missing values in the data. To be precise '5,970' records are duplicated records. We will drop these records from the dataset as well. 

In [12]:
# drop the duplicated records 
df_final = df_clean.drop_duplicates().copy()
df_final

product_name  price                  category  \
0       Solid Men Multicolor Track Pants  2,999  Clothing and Accessories   
1             Solid Men Blue Track Pants  1,499  Clothing and Accessories   
4      Solid Men Brown, Grey Track Pants  2,999  Clothing and Accessories   
5       Solid Men Multicolor Track Pants  4,499  Clothing and Accessories   
6        Solid Men Dark Blue Track Pants  1,499  Clothing and Accessories   
...                                  ...    ...                       ...   
29973      Solid Men V Neck Grey T-Shirt    599  Clothing and Accessories   
29985       Full Sleeve Solid Men Jacket  2,999  Clothing and Accessories   
29989     Striped Men Black Basic Shorts    999  Clothing and Accessories   
29990      Solid Men Grey Bermuda Shorts    999  Clothing and Accessories   
29992       Solid Men Black Night Shorts    799  Clothing and Accessories   

                                             description  rating  
0      Yorker trackpants made from 100% rich combed c...     3.9  
1      Yorker trackpants made from 100% rich combed c...     3.9  
4      Yorker trackpants made from 100% rich combed c...     3.9  
5      Yorker trackpants made from 100% rich combed c...     3.8  
6      Yorker trackpants made from 100% rich combed c...     3.8  
...                                                  ...     ...  
29973  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29985  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29989  OKANE endeavors to continue to be a leader in ...     3.8  
29990  ‘OKANE’ endeavors to continue to be a leader i...     3.8  
29992  ‘OKANE’ endeavors to continue to be a leader i...     3.8  

[10112 rows x 5 columns]

Our final dataset will incldue only '10,112' records. This cleaned dataset will be stored and used to provide product details for the catalog. 
 

In [ ]:
# turn the price into a valid price
df_final['price'] = df_final['price'].apply(lambda x: float(str(x).replace(',', '')))

# turn the dataframe to create a dictionary for each record
records = df_final.to_dict(orient='records')

# write the records as a json file 
with open('../src/product_details.json', 'w', encoding='utf-8') as file:
    json.dump(records, file, indent=2)